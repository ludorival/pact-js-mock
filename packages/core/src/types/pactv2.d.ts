declare namespace PactV2 {
    /**
 * Schema for a Pact file
 */
    export type PactFile = {
        consumer: Pacticipant;
        interactions: Interaction[];
        metadata?: Metadata;
        provider: Pacticipant;
        [property: string]: any;
    }

    export type Pacticipant = {
        name: string;
        [property: string]: any;
    }

    export type Interaction<TResponse = any, TRequest = any,> = {
        description: string;
        providerState?: string;
        request: Request<TRequest>;
        response: Response<TResponse>;
    }

    export type Request<T = any> = {
        body?: T;
        headers?: { [key: string]: any };
        matchingRules?: MatchingRules;
        method: Method;
        path: string;
        query?: string;
    }

    export type MatchingRules = {
    }

    export type Method = "connect" | "CONNECT" | "delete" | "DELETE" | "get" | "GET" | "head" | "HEAD" | "options" | "OPTIONS" | "post" | "POST" | "put" | "PUT" | "trace" | "TRACE";

    export type Response<T = any> = {
        body?: T;
        headers?: { [key: string]: any };
        matchingRules?: MatchingRules;
        status: number;
    }

    export type Metadata = {
        "pact-specification"?: PactSpecification;
        pactSpecification?: PactSpecificationClass;
        pactSpecificationVersion?: string;
        [property: string]: any;
    }

    export type PactSpecification = {
        version: string;
    }

    export type PactSpecificationClass = {
        version: string;
    }

}