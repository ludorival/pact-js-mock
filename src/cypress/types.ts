export type HeadersConfig = {
  includes?: string[]
  excludes?: string[]
}
export type HeaderType = Record<string, string | string[]> | undefined
