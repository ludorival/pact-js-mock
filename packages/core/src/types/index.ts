export * as PactV2 from './pactv2'
export * as PactV3 from './pactv3'
export * as PactV4 from './pactv4'
