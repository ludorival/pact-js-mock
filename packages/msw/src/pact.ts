import {
  Pact as BasePact,
  InteractionFor,
  MinimalInteraction,
  PactFile,
  PactV2,
  PactV3,
  PactV4,
  Version,
} from '@pact-mock-js/core'
import { omit } from 'lodash'
import {
  GraphQLContext,
  MockedRequest,
  ResponseResolver,
  ResponseTransformer,
  RestContext,
} from 'msw'
import {
  GraphQLPayload,
  GraphQLResponse,
  HeaderType,
  HeadersConfig,
} from 'types'
export { MinimalInteraction, PactV2, PactV3, PactV4 }

type Options = {
  headersConfig?: HeadersConfig
  basePath?: string
}

function buildResponse<T>(content: T, version: Version) {
  switch (version) {
    case '2.0.0':
      return {
        response: { status: 200, body: content },
      } as MinimalInteraction<PactV2.Interaction>
    case '3.0.0':
      return {
        response: { status: 200, body: content },
      } as MinimalInteraction<PactV3.Interaction>

    case '4.0.0':
      return {
        response: {
          status: 200,
          body: { content, contentType: 'application/json' },
        },
      } as MinimalInteraction<PactV4.Interaction>
  }
}

type Request = PactV2.Request | PactV3.Request | PactV4.Request
export class Pact<P extends PactFile> extends BasePact {
  private options: Options
  constructor(pact: Omit<P, 'interactions'>, options?: Options) {
    super(pact)
    this.options = { ...options }
  }

  toResolver<T extends object>(
    input: MinimalInteraction<InteractionFor<P, T>> | T,
    once?: boolean
  ): ResponseResolver<
    MockedRequest,
    GraphQLContext<GraphQLPayload> | RestContext
  > {
    const version = this.version
    return async (req, res, ctx) => {
      const interaction =
        'response' in input ? input : buildResponse(input, version)
      const transformers = await this.toTransformers(
        interaction as MinimalInteraction<InteractionFor<P, T>>,
        req,
        ctx
      )
      if (once) return res.once(...transformers)
      else return res(...transformers)
    }
  }
  async toTransformers(
    interaction: MinimalInteraction<InteractionFor<P>>,
    req: MockedRequest,
    context: GraphQLContext<GraphQLPayload> | RestContext
  ): Promise<ResponseTransformer[]> {
    await toRequest(req, this.options)
      .then((request) =>
        this.record({
          ...interaction,
          request,
        })
      )
      .then()
    const responseV4 = (interaction.response as { body?: { content: unknown } })
      .body?.content
      ? (interaction.response as PactV4.ResponseClass)
      : null
    const response = responseV4
      ? null
      : (interaction.response as PactV2.Response | PactV3.Response)
    const content = responseV4?.body?.content || response?.body
    const graphQLResponse = content?.errors
      ? (content as GraphQLResponse)
      : content?.data
      ? (content as GraphQLResponse)
      : null
    const responseStatus = responseV4?.status || response?.status || 200
    if (graphQLResponse?.errors && 'errors' in context) {
      return [context.errors(graphQLResponse.errors)]
    } else if (graphQLResponse?.data && 'data' in context) {
      return [context.data(graphQLResponse.data)]
    } else if ('json' in context) {
      return [
        context.status(responseStatus),
        ...(content && (typeof content === 'object' || Array.isArray(content))
          ? [context.json(content)]
          : content
          ? [context.body(content.toString())]
          : []),
      ]
    } else {
      return [context.status(responseStatus)]
    }
  }
}

async function toRequest(
  req: MockedRequest,
  { headersConfig, basePath }: Options
): Promise<Request> {
  const path = req.url.pathname.replace(basePath || '', '')
  const query = req.url.searchParams.toString() || undefined
  const body: Body = await req
    .json()
    .catch(() => req.text())
    .catch(() => undefined)

  return {
    method: req.method,
    path,
    headers: omitHeaders(req.headers.all(), headersConfig),
    body,
    query,
  } as Request
}

const omitHeaders = (
  headers: HeaderType,
  headersConfig: HeadersConfig = {}
) => {
  const blocklist = headersConfig.excludes || []
  if (headersConfig.includes) {
    const remove = Object.keys(omit(headers, ...headersConfig.includes))
    blocklist.push(...remove)
  }
  return omit(headers, [...blocklist])
}
