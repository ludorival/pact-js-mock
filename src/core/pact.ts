import { omit } from 'lodash'
import {
  HeaderType,
  HeadersConfig,
  InputPact,
  InteractionFor,
  MinimalInteraction,
  Options,
  PactFile,
  PactV2,
  PactV3,
  PactV4,
  Request,
  ToRecordInteraction,
  Version,
} from '../types'
import packageJson from '../../package.json'

function pactName<P extends PactFile>(pact: InputPact<P>) {
  return `${pact.consumer.name}-${pact.provider.name}`
}
export class Pact<T extends PactFile = PactV2.PactFile> {
  private interactions: Record<string, InteractionFor<T>> = {}
  constructor(
    private pact: InputPact<T>,
    private options?: Options,
  ) {
    if (!pact.metadata?.pactSpecification?.version)
      throw new Error(`The version is missing in the Pact. Please provide the right version like
    { consumer: { name : 'my-consumer'}, provider : { name: 'my-provider'},
    metadata : {pactSpecification: { version : '2.0.0'}}}`)
  }

  public get name(): string {
    return pactName(this.pact)
  }
  public get version(): Version {
    return this.pact.metadata?.pactSpecification?.version || '2.0.0'
  }

  record<TResponse = unknown, TRequest = unknown>(
    input: ToRecordInteraction<InteractionFor<T, TResponse, TRequest>>,
  ) {
    const request = input.request as Request
    const response = input.response as { status?: number }
    const description =
      input.description ||
      `${input.request?.method} ${input.request?.path} returns status ${
        response.status || 200
      }`

    const interaction = {
      description,
      ...input,
      request: { ...this.toRequest(request) },
    } as InteractionFor<T, TResponse, TRequest>
    const existingInteraction = this.interactions[description]
    if (
      existingInteraction &&
      JSON.stringify(existingInteraction) !== JSON.stringify(interaction)
    ) {
      console.warn(
        `The interaction ${description} already exists but with different content. It is recommended that the interaction stays deterministic.`,
      )
    }
    this.interactions[description] = interaction
  }

  private toRequest(req: Request): Request {
    const { headersConfig, basePath } = this.options || {}
    const path = req.path?.replace(basePath || '', '')
    if (req.headers) {
      req.headers = omitHeaders(req.headers, headersConfig)
    }
    return {
      ...req,
      path,
    } as Request
  }

  generatePactFile(): T {
    const interactions = Object.keys(this.interactions)
      .map((d) => this.interactions[d])
      .sort((a, b) => a.description.localeCompare(b.description))
    const pactFile = {
      ...(this.pact as T),
      interactions,
      metadata: {
        ...this.pact.metadata,
        client: { name: packageJson.name, version: packageJson.version },
      },
    }
    return pactFile
  }

  async reset() {
    this.interactions = {}
  }
}

const omitHeaders = (
  headers: HeaderType,
  headersConfig: HeadersConfig = {},
) => {
  const blocklist = headersConfig.excludes || []
  if (headersConfig.includes) {
    const remove = Object.keys(omit(headers, ...headersConfig.includes))
    blocklist.push(...remove)
  }
  return omit(headers, [...blocklist])
}

export function buildResponse<T>(content: T, version: Version) {
  switch (version) {
    case '2.0.0':
      return {
        response: { status: 200, body: content },
      } as MinimalInteraction<PactV2.Interaction>
    case '3.0.0':
      return {
        response: { status: 200, body: content },
      } as MinimalInteraction<PactV3.Interaction>

    case '4.0.0':
      return {
        response: {
          status: 200,
          body: { content, contentType: 'application/json' },
        },
      } as MinimalInteraction<PactV4.Interaction>
  }
}
