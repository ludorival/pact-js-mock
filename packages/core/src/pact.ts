import { mkdir, readFile, readdir, writeFile, rm } from 'fs/promises'
import { existsSync } from 'fs'
import { join, resolve } from 'path'
export * from './types'
import { PactV2, PactV3, PactV4 } from 'types'

export type PactFile<I> = I extends PactV2.Interaction
  ? PactV2.PactFile
  : I extends PactV3.Interaction
  ? PactV3.PactFile
  : PactV4.PactFile

export type Interaction<TResponse = any> =
  | PactV2.Interaction<TResponse>
  | PactV3.Interaction<TResponse>
  | PactV4.Interaction<TResponse>

export type MinimalInteraction<I extends Interaction = PactV2.Interaction> =
  Partial<Omit<I, 'response'>> & Pick<I, 'response'>

export type ToRecordInteraction<I extends Interaction> = Partial<
  Omit<I, 'response' | 'request'>
> &
  Pick<I, 'response' | 'request'>
function sanitizeFileName(input: string): string {
  return input
    .toLowerCase()
    .trim()
    .replace(/[^\w\s-]/g, '')
    .replace(/\s+/g, '-')
    .replace(/-+/g, '-')
    .replace(/^[-_]+|[-_]+$/g, '')
}

function pactName<I>(pact: Omit<PactFile<I>, 'interactions'>) {
  return `${pact.consumer.name}-${pact.provider.name}`
}
export class Pact<I extends Interaction = PactV2.Interaction> {
  private outputDir: string
  constructor(
    private pact: Omit<PactFile<I>, 'interactions'>,
    outputDir = 'pacts'
  ) {
    this.outputDir = resolve(join(outputDir, pactName<I>(pact)))
    mkdir(this.outputDir, { recursive: true }).catch((e) =>
      console.error(
        `There is an error when trying to create the pact output directory: ${e}`
      )
    )
  }

  public get name(): string {
    return pactName(this.pact)
  }
  public get version(): string {
    return this.pact.metaData.pactSpecification?.version || '2.0.0'
  }

  async record<T = any>(interaction: ToRecordInteraction<I>): Promise<T> {
    const response = interaction.response as { status?: number }
    const description =
      interaction.description ||
      `${interaction.request?.method} ${
        interaction.request?.path
      } returns status ${response.status || 200}`

    const fileName = join(
      this.outputDir,
      `${sanitizeFileName(description)}.interaction.json`
    )
    if (existsSync(fileName)) {
      console.warn(
        `The file ${resolve(fileName)} already exists and will be overriden`
      )
    }
    await mkdir(this.outputDir, { recursive: true })
    await writeFile(fileName, JSON.stringify({ description, ...interaction }))
    if (
      typeof interaction.response === 'object' &&
      interaction.response &&
      'status' in interaction.response
    ) {
      return interaction.response.body
    }
    if (Array.isArray(interaction.response)) {
      return interaction.response[0]?.contents
    }
    return interaction.response as T
  }

  async generatePactFile(): Promise<PactFile<I>> {
    const interactionFiles = (await readdir(this.outputDir)).filter((f) =>
      f.endsWith('interaction.json')
    )
    const interactions: Interaction[] = await Promise.all(
      interactionFiles.map((f) =>
        readFile(join(this.outputDir, f)).then(
          (buffer) => JSON.parse(buffer.toString()) as Interaction
        )
      )
    )
    const pactFile = { ...this.pact, interactions } as PactFile<I>
    await writeFile(
      join(this.outputDir, '..', `${this.name}.json`),
      JSON.stringify(pactFile, null, 2)
    )
    Promise.all(interactionFiles.map((f) => rm(join(this.outputDir, f))))
      .catch((e) =>
        console.warn(
          `Got an error when deleting an interaction file in '${this.outputDir}' : ${e}`
        )
      )
      .then()
    return pactFile
  }

  public get fileName(): string {
    return join(this.outputDir, '..', `${this.name}.json`)
  }

  async reset() {
    await rm(this.outputDir, { recursive: true }).catch((e) =>
      console.warn(
        `Cannot remove the pact directory output '${this.outputDir}' : ${e}`
      )
    )
  }
}
