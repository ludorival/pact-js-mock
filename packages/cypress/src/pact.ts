import type {
  CyHttpMessages,
  RouteHandlerController,
  StaticResponse,
} from 'cypress/types/net-stubbing'
import { omit } from 'lodash'
import {
  Pact as BasePact,
  InputPact,
  InteractionFor,
  MinimalInteraction,
  PactFile,
  PactV2,
  PactV3,
  PactV4,
  Version,
} from 'pact-mock-js.core'
import { HeaderType, HeadersConfig } from './types'

export { MinimalInteraction, PactV2, PactV3, PactV4 }

type Options = {
  headersConfig?: HeadersConfig
  basePath?: string
}

function buildResponse<T>(content: T, version: Version) {
  switch (version) {
    case '2.0.0':
      return {
        response: { status: 200, body: content },
      } as MinimalInteraction<PactV2.Interaction>
    case '3.0.0':
      return {
        response: { status: 200, body: content },
      } as MinimalInteraction<PactV3.Interaction>

    case '4.0.0':
      return {
        response: {
          status: 200,
          body: { content, contentType: 'application/json' },
        },
      } as MinimalInteraction<PactV4.Interaction>
  }
}

type Request = PactV2.Request | PactV3.Request | PactV4.Request
export class Pact<P extends PactFile> extends BasePact {
  private options: Options
  constructor(pact: InputPact<P>, options?: Options) {
    super(pact as any)
    this.options = { ...options }
  }

  toHandler<T extends object>(
    input: MinimalInteraction<InteractionFor<P, T>> | T
  ): RouteHandlerController {
    const version = this.version
    return (req) => {
      req.reply()
      const interaction =
        'response' in input ? input : buildResponse(input, version)
      return req.reply(
        this.recordResponse(
          interaction as MinimalInteraction<InteractionFor<P, T>>,
          req
        )
      )
    }
  }
  recordResponse(
    interaction: MinimalInteraction<InteractionFor<P>>,
    req: CyHttpMessages.IncomingHttpRequest
  ): StaticResponse {
    const request = toRequest(req, this.options)
    this.record({ ...interaction, request } as any)
    const responseV4 = (interaction.response as { body?: { content: unknown } })
      .body?.content
      ? (interaction.response as PactV4.ResponseClass)
      : null
    const response = responseV4
      ? null
      : (interaction.response as PactV2.Response | PactV3.Response)

    return {
      statusCode: responseV4?.status || response?.status,
      body: responseV4?.body || response?.body,
      headers: responseV4?.headers || response?.headers,
    }
  }
}

function toRequest(
  req: CyHttpMessages.IncomingHttpRequest,
  { headersConfig, basePath }: Options
): Request {
  const parts = req.url.split('?')
  const path = parts[0].replace(basePath || '', '')
  const query = parts.slice(1).join('?')
  const body: Body = req.body

  return {
    method: req.method,
    path,
    headers: omitHeaders(req.headers, headersConfig),
    body,
    query,
  } as Request
}

const omitHeaders = (
  headers: HeaderType,
  headersConfig: HeadersConfig = {}
) => {
  const blocklist = headersConfig.excludes || []
  if (headersConfig.includes) {
    const remove = Object.keys(omit(headers, ...headersConfig.includes))
    blocklist.push(...remove)
  }
  return omit(headers, [...blocklist])
}
