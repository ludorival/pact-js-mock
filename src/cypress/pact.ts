import type {
  CyHttpMessages,
  RouteHandlerController,
  StaticResponse,
} from 'cypress/types/net-stubbing'
import { Pact as BasePact, buildResponse } from '../core'
import {
  InputPact,
  InteractionFor,
  MinimalInteraction,
  PactFile,
  PactV2,
  PactV3,
  PactV4,
  ToRecordInteraction,
} from '../types'
import { HeadersConfig } from './types'

type Options = {
  headersConfig?: HeadersConfig
  basePath?: string
}

type Request = PactV2.Request | PactV3.Request | PactV4.Request
export class Pact<P extends PactFile> extends BasePact<P> {
  constructor(pact: InputPact<P>, options?: Options) {
    super(pact, options)
  }

  toHandler<T extends object>(
    input: MinimalInteraction<InteractionFor<P, T>> | T
  ): RouteHandlerController {
    const version = this.version
    return (req) => {
      const interaction =
        'response' in input ? input : buildResponse(input, version)
      req.reply(
        this.recordResponse(
          interaction as MinimalInteraction<InteractionFor<P, T>>,
          req
        )
      )
    }
  }
  recordResponse(
    interaction: MinimalInteraction<InteractionFor<P>>,
    req: CyHttpMessages.IncomingHttpRequest
  ): StaticResponse {
    const request = toRequest(req)
    this.record({ ...interaction, request } as ToRecordInteraction<
      InteractionFor<P>
    >)
    const responseV4 = (interaction.response as { body?: { content: unknown } })
      .body?.content
      ? (interaction.response as PactV4.ResponseClass)
      : null
    const response = responseV4
      ? null
      : (interaction.response as PactV2.Response | PactV3.Response)

    return {
      statusCode: responseV4?.status || response?.status,
      body: responseV4?.body || response?.body,
      headers: responseV4?.headers || response?.headers,
    }
  }
}

function toRequest(req: CyHttpMessages.IncomingHttpRequest): Request {
  const parts = req.url.split('?')
  const path = parts[0]
  const query = parts.slice(1).join('?')
  const body: Body = req.body

  return {
    method: req.method,
    path,
    headers: req.headers,
    body,
    query,
  } as Request
}
