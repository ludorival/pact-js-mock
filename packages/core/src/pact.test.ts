import { PactV2, PactV3, PactV4 } from 'types'
import { Pact } from './pact'
import { readFile } from 'fs/promises'
describe('PactV2', () => {
  const pact = new Pact(
    {
      consumer: { name: 'consumer' },
      provider: { name: 'provider' },
      metadata: { pactSpecification: { version: '2.0.0' } },
    },
    'pacts/v2'
  )

  afterEach(async () => {
    await pact.reset()
  })

  it('should record a Pact V2 Interactions ', async () => {
    // given

    const anInteraction = {
      description: 'This is a description',
      request: { method: 'POST', path: 'v1/todo', body: { name: 'Todo' } },
      response: {
        status: 200,
        body: 'This is an interaction with a description',
      },
    } as PactV2.Interaction

    // when
    await pact.record<string>({
      request: { method: 'GET', path: 'v1/todo' },
      response: {
        status: 200,
        body: 'This is an interaction without description',
      },
    })
    await pact.record(anInteraction)
    await pact.record({
      ...anInteraction,
      description: 'This is a same description',
    })
    await pact.record({
      ...anInteraction,
      description: 'This is a same description',
      request: { method: 'GET', path: 'v1/todo' },
    })
    await pact.record({
      ...anInteraction,
      description: 'Description with special character @/"". $',
    })
    await pact.record({
      ...anInteraction,
      description: 'This is an interaction with matching rules',
      request: {
        ...anInteraction.request,
        matchingRules: { '$.body': { match: 'type' } },
      },
    })

    const pactFile = await pact.generatePactFile()
    expect(pactFile).toMatchSnapshot()
    await readFile(pact.fileName)
  })
})

describe('PactV3', () => {
  const pact = new Pact<PactV3.Interaction>(
    {
      consumer: { name: 'consumer' },
      provider: { name: 'provider' },
      metadata: { pactSpecification: { version: '3.0.0' } },
    },
    'pacts/v3'
  )

  afterEach(async () => {
    await pact.reset()
  })

  it('should record a Pact V3 Interactions ', async () => {
    // given

    const anInteraction = {
      description: 'This is a description',
      request: { method: 'POST', path: 'v1/todo', body: { name: 'Todo' } },
      response: {
        status: 200,
        body: 'This is an interaction with a description',
      },
    } as PactV3.Interaction

    // when
    await pact.record<string>({
      request: { method: 'GET', path: 'v1/todo' },
      response: {
        status: 200,
        body: 'This is an interaction without description',
      },
    })
    await pact.record(anInteraction)
    await pact.record({
      ...anInteraction,
      description: 'This is a same description',
    })
    await pact.record({
      ...anInteraction,
      description: 'This is a same description',
      request: { method: 'GET', path: 'v1/todo' },
    })
    await pact.record({
      ...anInteraction,
      description: 'Description with special character @/"". $',
    })
    await pact.record({
      ...anInteraction,
      description: 'This is an interaction with matching rules',
      request: {
        ...anInteraction.request,
        matchingRules: { body: { match: 'type' } },
      },
    })

    const pactFile = await pact.generatePactFile()
    expect(pactFile).toMatchSnapshot()
    await readFile(pact.fileName)
  })
})

describe('PactV4', () => {
  const pact = new Pact<PactV4.Interaction>(
    {
      consumer: { name: 'consumer' },
      provider: { name: 'provider' },
      metadata: { pactSpecification: { version: '4.0.0' } },
    },
    'pacts/v4'
  )

  afterEach(async () => {
    await pact.reset()
  })

  it('should record a Pact V4 Interactions ', async () => {
    // given

    const anInteraction = {
      type: 'Synchronous/HTTP',
      description: 'This is a description',
      request: {
        method: 'POST',
        path: 'v1/todo',
        body: { content: { name: 'Todo' }, contentType: 'application/json' },
      },
      response: {
        body: {
          content: 'This is an interaction with a description',
          contentType: 'application/json',
        },
      },
    } as PactV4.Interaction

    // when
    await pact.record<string>({
      request: { method: 'GET', path: 'v1/todo' },
      response: {
        status: 200,
        body: {
          content: 'This is an interaction without description',
          contentType: 'application/json',
        },
      },
    })
    await pact.record(anInteraction)
    await pact.record({
      ...anInteraction,
      description: 'This is a same description',
    })
    await pact.record({
      ...anInteraction,
      description: 'This is a same description',
      request: { method: 'GET', path: 'v1/todo' },
    })
    await pact.record({
      ...anInteraction,
      description: 'Description with special character @/"". $',
    })
    await pact.record({
      ...anInteraction,
      description: 'This is an interaction with matching rules',
      request: {
        ...anInteraction.request,
        matchingRules: { body: { match: 'type' } },
      },
    })

    const pactFile = await pact.generatePactFile()
    expect(pactFile).toMatchSnapshot()
    await readFile(pact.fileName)
  })
})
