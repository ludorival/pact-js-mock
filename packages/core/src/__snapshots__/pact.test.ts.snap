// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PactV2 should record a Pact V2 Interactions  1`] = `
{
  "consumer": {
    "name": "consumer",
  },
  "interactions": [
    {
      "description": "Description with special character @/"". $",
      "request": {
        "body": {
          "name": "Todo",
        },
        "method": "POST",
        "path": "v1/todo",
      },
      "response": {
        "body": "This is an interaction with a description",
        "status": 200,
      },
    },
    {
      "description": "GET v1/todo returns status 200",
      "request": {
        "method": "GET",
        "path": "v1/todo",
      },
      "response": {
        "body": "This is an interaction without description",
        "status": 200,
      },
    },
    {
      "description": "This is a description",
      "request": {
        "body": {
          "name": "Todo",
        },
        "method": "POST",
        "path": "v1/todo",
      },
      "response": {
        "body": "This is an interaction with a description",
        "status": 200,
      },
    },
    {
      "description": "This is a same description",
      "request": {
        "method": "GET",
        "path": "v1/todo",
      },
      "response": {
        "body": "This is an interaction with a description",
        "status": 200,
      },
    },
    {
      "description": "This is an interaction with matching rules",
      "request": {
        "body": {
          "name": "Todo",
        },
        "matchingRules": {
          "$.body": {
            "match": "type",
          },
        },
        "method": "POST",
        "path": "v1/todo",
      },
      "response": {
        "body": "This is an interaction with a description",
        "status": 200,
      },
    },
  ],
  "metadata": {
    "pactSpecification": {
      "version": "2.0.0",
    },
  },
  "provider": {
    "name": "provider",
  },
}
`;

exports[`PactV3 should record a Pact V3 Interactions  1`] = `
{
  "consumer": {
    "name": "consumer",
  },
  "interactions": [
    {
      "description": "Description with special character @/"". $",
      "request": {
        "body": {
          "name": "Todo",
        },
        "method": "POST",
        "path": "v1/todo",
      },
      "response": {
        "body": "This is an interaction with a description",
        "status": 200,
      },
    },
    {
      "description": "GET v1/todo returns status 200",
      "request": {
        "method": "GET",
        "path": "v1/todo",
      },
      "response": {
        "body": "This is an interaction without description",
        "status": 200,
      },
    },
    {
      "description": "This is a description",
      "request": {
        "body": {
          "name": "Todo",
        },
        "method": "POST",
        "path": "v1/todo",
      },
      "response": {
        "body": "This is an interaction with a description",
        "status": 200,
      },
    },
    {
      "description": "This is a same description",
      "request": {
        "method": "GET",
        "path": "v1/todo",
      },
      "response": {
        "body": "This is an interaction with a description",
        "status": 200,
      },
    },
    {
      "description": "This is an interaction with matching rules",
      "request": {
        "body": {
          "name": "Todo",
        },
        "matchingRules": {
          "body": {
            "match": "type",
          },
        },
        "method": "POST",
        "path": "v1/todo",
      },
      "response": {
        "body": "This is an interaction with a description",
        "status": 200,
      },
    },
  ],
  "metadata": {
    "pactSpecification": {
      "version": "3.0.0",
    },
  },
  "provider": {
    "name": "provider",
  },
}
`;

exports[`PactV4 should record a Pact V4 Interactions  1`] = `
{
  "consumer": {
    "name": "consumer",
  },
  "interactions": [
    {
      "description": "Description with special character @/"". $",
      "request": {
        "body": {
          "content": {
            "name": "Todo",
          },
          "contentType": "application/json",
        },
        "method": "POST",
        "path": "v1/todo",
      },
      "response": {
        "body": {
          "content": "This is an interaction with a description",
          "contentType": "application/json",
        },
      },
      "type": "Synchronous/HTTP",
    },
    {
      "description": "GET v1/todo returns status 200",
      "request": {
        "method": "GET",
        "path": "v1/todo",
      },
      "response": {
        "body": {
          "content": "This is an interaction without description",
          "contentType": "application/json",
        },
        "status": 200,
      },
    },
    {
      "description": "This is a description",
      "request": {
        "body": {
          "content": {
            "name": "Todo",
          },
          "contentType": "application/json",
        },
        "method": "POST",
        "path": "v1/todo",
      },
      "response": {
        "body": {
          "content": "This is an interaction with a description",
          "contentType": "application/json",
        },
      },
      "type": "Synchronous/HTTP",
    },
    {
      "description": "This is a same description",
      "request": {
        "method": "GET",
        "path": "v1/todo",
      },
      "response": {
        "body": {
          "content": "This is an interaction with a description",
          "contentType": "application/json",
        },
      },
      "type": "Synchronous/HTTP",
    },
    {
      "description": "This is an interaction with matching rules",
      "request": {
        "body": {
          "content": {
            "name": "Todo",
          },
          "contentType": "application/json",
        },
        "matchingRules": {
          "body": {
            "match": "type",
          },
        },
        "method": "POST",
        "path": "v1/todo",
      },
      "response": {
        "body": {
          "content": "This is an interaction with a description",
          "contentType": "application/json",
        },
      },
      "type": "Synchronous/HTTP",
    },
  ],
  "metadata": {
    "pactSpecification": {
      "version": "4.0.0",
    },
  },
  "provider": {
    "name": "provider",
  },
}
`;
